core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala

```diff
@@ -0,0 +1,251 @@
+     * is kept and updated upon each invocation.
+     *
+     * If asked, check for sentinel markers - but exclude the identified markers from
+     * the result returned from this method.
+     *
+     * Only parses and returns as much logs as fit in the passed log limit.
+     * Even if the log limit is exceeded, advance the starting position for the next invocation
+     * behind the bytes most recently read - but don't actively read any more until sentinel
+     * markers have been found.
+     *
+     * @param limit the limit to apply to the log size
+     * @param waitForSentinel determines if the processor should wait for a sentinel to appear
+     *
+     * @return a vector of Strings with log lines in our own JSON format
+     */
+    def logs(limit: ByteSize, waitForSentinel: Boolean)(implicit transid: TransactionId): Future[Vector[String]] = {
+
+        def readLogs(retries: Int): Future[Vector[String]] = {
+            docker.rawContainerLogs(id, logFileOffset).flatMap { rawLogBytes =>
+                val rawLog = new String(rawLogBytes.array, rawLogBytes.arrayOffset, rawLogBytes.position, StandardCharsets.UTF_8)
+                val (isComplete, isTruncated, formattedLogs) = processJsonDriverLogContents(rawLog, waitForSentinel, limit)
+
+                if (retries > 0 && !isComplete && !isTruncated) {
+                    log.info(this, s"log cursor advanced but missing sentinel, trying $retries more times")
+                    Thread.sleep(logsRetryWait.toMillis)
+                    readLogs(retries - 1)
+                } else {
+                    logFileOffset += rawLogBytes.position - rawLogBytes.arrayOffset
+                    Future.successful(formattedLogs)
+                }
+            }.andThen {
+                case Failure(e) =>
+                    log.error(this, s"Failed to obtain logs of ${id.asString}: ${e.getClass} - ${e.getMessage}")
+            }
+        }
+
+        readLogs(logsRetryCount)
+    }
+
+    /**
+     * Makes an HTTP request to the container.
+     *
+     * Note that `http.post` will not throw an exception, hence the generated Future cannot fail.
+     *
+     * @param path relative path to use in the http request
+     * @param body body to send
+     * @param timeout timeout of the request
+     * @param retry whether or not to retry the request
+     */
+    protected def callContainer(path: String, body: JsObject, timeout: FiniteDuration, retry: Boolean = false): Future[RunResult] = {
+        val started = Instant.now()
+        val http = httpConnection.getOrElse {
+            val conn = new HttpUtils(s"${ip.asString}:8080", timeout, 1.MB)
+            httpConnection = Some(conn)
+            conn
+        }
+        Future {
+            http.post(path, body, retry)
+        }.map { response =>
+            val finished = Instant.now()
+            RunResult(Interval(started, finished), response)
+        }
```

Comment:
I'm not sure what you have in mind. At the moment, we have two consumers of `callContainer()`, i.e. `initialize()` and `run()`. Those two consumers deal with a failed `Future`.

If we add a `recover` here and return a successful `Future[RunResult]` in any case, the caller's error handling isn't used any more.

So should we keep the error handling as is - or move it down to `callContainer()`?